wallet/src/Pos/Util/Mnemonic.hs

```diff
@@ -0,0 +1,683 @@
+-- | Module providing restoring from backup phrase functionality
+{-# LANGUAGE DataKinds      #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE TypeFamilies   #-}
+{-# OPTIONS_GHC -Wno-redundant-constraints #-}
+
+module Pos.Util.Mnemonic
+       (
+       -- * Types
+         Mnemonic
+       , Entropy
+       , SupportedWords
+
+       -- * Creating @Mnemonic@ (resp. @Entropy@)
+       , mkEntropy
+       , mkMnemonic
+       , genEntropy
+
+       -- * Access properties of types
+       , entropyLength
+       , checksumLength
+
+       -- * Converting from and to @Mnemonic@ (resp. @Entropy@)
+       , mnemonicToEntropy
+       , mnemonicToSeed
+       , mnemonicToAESKey
+       , entropyToMnemonic
+       , entropyToByteString
+       ) where
+
+import           Universum
+
+import           Control.Lens ((?~))
+import           Crypto.Hash (Blake2b_256, Digest, SHA256, hash)
+import           Data.Aeson (FromJSON (..), ToJSON (..))
+import           Data.Aeson.Types (Parser)
+import           Data.ByteArray (convert)
+import           Data.ByteString (ByteString)
+import           Data.Char (isAscii)
+import           Data.Default (Default (def))
+import           Data.List (elemIndex, (!!))
+import           Data.Swagger (NamedSchema (..), ToSchema (..), maxItems, minItems)
+import           Data.Text.Buildable (Buildable (build))
+import           GHC.TypeLits (ErrorMessage (..), TypeError)
+import           Test.QuickCheck (Arbitrary (..))
+import           Test.QuickCheck.Gen (vectorOf)
+
+import           Pos.Binary (serialize')
+import           Pos.Crypto (AesKey (..))
+import           Pos.Util.Bits (Bit, Word11, fromBits, toBits)
+import           Pos.Util.LogSafe (SecureLog)
+
+
+import qualified Crypto.Random.Entropy as Crypto
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Char8 as B8
+import qualified Data.Text as Text
+
+
+-- | A backup-phrase in the form of a non-empty of Mnemonic words
+-- Constructor isn't exposed.
+data Mnemonic (n :: Nat) = Mnemonic
+    { getEntropy :: Entropy n
+    , getWords   :: [(Word11, Text)]
+    } deriving (Eq, Show, Generic)
+
+-- | Entropy as a non-empty sequence of bytes, multiple of 4 bytes
+newtype Entropy (n :: Nat) = Entropy ByteString deriving (Eq, Show, Generic)
+
+type family SupportedWords (n :: Nat) :: Constraint where
+    SupportedWords 3  = ()
+    SupportedWords 6  = ()
+    SupportedWords 9  = ()
+    SupportedWords 12 = ()
+    SupportedWords 15 = ()
+    SupportedWords 18 = ()
+    SupportedWords 21 = ()
+    SupportedWords 24 = ()
+    SupportedWords 27 = ()
+    SupportedWords 30 = ()
+    SupportedWords 33 = ()
+    SupportedWords 36 = ()
+    SupportedWords 39 = ()
+    SupportedWords 42 = ()
+    SupportedWords 45 = ()
+    SupportedWords 48 = ()
+    SupportedWords _  = TypeError ('Text "Unsupported number of Mnemonic words")
```

Comment:
I feel bad about this. _In theory_, to maintain 100% compatibility with the current implementation, we have to support this whole list. However, I don't think we used any mnemonics but with 9 and 12 words. 
@akegalj May you confirm / disconfirm this?

Reply:
right. To my knowledge we only use 9 and 12 word mnemonics. I am not sure about recently developed paper certificates. Maybe @darko-mijic can fill in the gaps.